File: solitaire.pl
#!/usr/bin/perl -l

#NOTE: THIS DOES NOT USE ANY REFERENCES OR MOOSE OR ANYTHING LIKE THAT, IT PASSES EVERYTHING AROUND AS LIST LITERALS BEACAUSE 
#WELL, I DIDN'T REALLY CARE IF THEY WEREN'T PASSED AROUND EFFICIENTLY. IF YOU ARE SOMEONE TRYING TO LEARN FROM THIS SCRIPT, 
#I AM VERY SORRY. THAT IS ALL.

#NOTE: IF YOU WANT DEBUGGING INFORMATION, SET THE ENVIRONMENT VARIABLE DEBUG=1!

#ALSO, I HAVE NOW USED ONE REFERENCE. THIS IS STILL NOT PRETTY CODE.

use strict;
use warnings;

my $DEBUG = $ENV{DEBUG};

#suits
my @suits = qw/D C H S/;

#ranks
my @ranks = qw/2 3 4 5 6 7 8 9 10 J Q K A/;

#seed random:
srand(rand(time()));

##MAIN FUNCTION:

#create a deck
my @deck = create_deck();

my @shuffled_deck = shuffle_deck(@deck);

#create a playing field, it should have 7 slots. Each slot should have a reserve and an active card.
#when the active card is played, a card is drawn from the reserve, and if there is no reserve, the
#slot is marked as empty


my $board_length = 7;
#deal out the board:
my @board;

#filling it out the easy way:
for(my $i=0; $i<$board_length; $i++) {
    my @row;
    for(my $j = $i; $j < $board_length; $j++) {
        push(@row, pop(@shuffled_deck));
    }
    push(@board, \@row);
}

#there is also a main reserve pile, of the cards left over after dealing.
#for now, this is just the deck.








#print the debug info:
print_debug();

##END MAIN FUNCTION

##FIELD FUNCTIONS



##END FIELD FUNCTIONS


##DECK FUNCTIONS
sub create_deck {
    my @deck;
    foreach my $rank(@ranks) {
        foreach my $suit(@suits) {
            push(@deck, "$rank$suit");
        }
    }
    #TODO: maybe use a ref here?
    return @deck;
}

sub shuffle_deck {
#shuffle the deck

    my @retval = @_;

# [edited notation]
# from wikipedia: http://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
    for(my $i=scalar(@retval) - 1; $i>0 ; $i--) {
        my $j = rand($i);
        @retval[$i, $j] = @retval[$j, $i];
    }
    return @retval;
}

##END DECK FUNCTIONS

##DEBUG FUNCTIONS

#debug / test function
sub print_debug {
    if($DEBUG) {
        print "suits:";
        print @suits;
        print "rank:";
        print @ranks;
        print "deck:";
        print @deck;
        print @shuffled_deck;
        print "board:";
#        print @board;

        foreach my $row(reverse @board) {
            my $column_out = join(" ", @$row);
            print $column_out;
        }
        
        
        print "reserve:";
        print join(" ", @deck);
    }
}

File: solitaire.rb
#!/usr/bin/ruby -w


class Card
  attr_reader :suit, :rank
  def initialize(suit,rank)
    @suit = suit
    @rank = rank
  end

  def to_s()
    return "#{rank} of #{suit}"
  end
end

class Deck
  @@suits = [:diamond, :clubs, :hearts, :spades]
  @@ranks = [:two, :three, :four, :five, :six, :seven, :eight, :nine, :ten, :jack, :queen, :king, :ace]
  attr_accessor :deck
  def initialize()
    @deck = Array.new
    @@suits.each do |suit|
      @@ranks.each do |rank|
        @deck.push(Card.new(suit,rank));
      end
    end
    puts "shuffling"
    shuffle()
  end

  def shuffle()
    @deck = @deck.sort_by { rand } 
  end

  def deal()
    return @deck.pop
  end

  def reserve()
    return @deck.length
  end
end

class Board
  attr_accessor :grid

  def initialize(deck)
    #clear out the right sizes for each array in the board.
    @grid = Array.new();
    for i in (1..7)
      @grid[i-1] = Array.new(i)      
    end
    
    #set up from the deck
    for i in (0.upto(@grid.length-1))
      for j in (0.upto(@grid[i].length-1))
        @grid[i][j] = deck.deal        
      end
    end
  end

end


class Game
  attr_accessor :board, :reserve, :foundations
  
  def initialize()
    @foundations = {:aces => [], :diamonds => [], :spades => [], :clubs => []}
    @reserve = Deck.new
    @board = Board.new(@reserve)
  end

  def play()
    while(0 != @reserve.reserve()) do
      puts @reserve.deal()
    end
  end

end

##MAIN

game = Game.new()

game.play()
